src/broker/overlay.c

```diff
@@ -1542,6 +1542,20 @@ const char *overlay_get_subtree_status (struct overlay *ov, int rank)
     return result;
 }
 
+struct idset *overlay_get_critical_ranks (struct overlay *ov)
+{
+    struct idset *ranks;
+
+    /* For now, return all internal ranks plus rank 0
+     */
+    if (!(ranks = topology_get_internal_ranks (ov->topo))
+        || idset_set (ranks, 0) < 0) {
+        idset_destroy (ranks);
+        return NULL;
+    }
+    return ranks;
+}
+
```

Comment:
Suggestion: take "critical" out of the name of this function since "what's critical" is to be configurable by the caller.  Maybe `overlay_get_nonleaf_ranks()`?

Reply:
Good point, but the function will always return rank 0 even for the degenerate case of an instance of size=1. This is why I added `critical` instead of nonleaf or internal. Any other suggestions?

Reply:
I'm not sure what to call this function, reading up a bit the terminology seems to indicate that a tree of size=1 is a leaf by itself, so I hesitate to use the term "nonleaf" ranks since that is technically incorrect.

Here's my thinking on keeping the name "critical". The set of critical nodes is a property of the overlay, if a "caller" or user of the overlay wants to change what is in the critical set then that should be a separate method of the overlay. We may also eventually allow the overlay to be hand-configured, in which case the critical set could be part of the configuration. In these cases, we still need a getter to return the critical set of ranks so could this function stay as-is and and perhaps the implementation changes based on the underlying overlay object properties?

Reply:
How about `overlay_get_default_critical_ranks()`?  Then it communicates that these are not the final critical ranks.  However, this is an internal function so if you still prefer the original, I'm OK with that!

Reply:
I'm confused - where are the "final critical ranks"? I'm missing something...

Reply:
Isn't this providing the default setting for the `broker.critical-ranks` attribute, which can be overridden on the command line?

Reply:
Ah, ok, yes. But I was considering the `struct overlay` as a class which knows nothing about broker attributes, and the critical-ranks idset as an attribute of an instance of that class. The broker then queries the overlay for what it considers critical ranks if the attribute isn't set on the command line. I guess calling it "default" makes an assumption it is only being used for the one purpose, but that is true so I'm fine changing the name.

Reply:
Ok, I changed the name of the function, squashed and force-pushed the result.

Reply:
Meh, my suggestion was probably lame.  You pick.

Reply:
OMG, this was my fault. I was somehow looking at the wrong code (the code added to overlay.c), so I apologize for all my confusing nonsense!